docs/architecture/adr-024-coin-metadata.md

```diff
@@ -0,0 +1,104 @@
+# ADR 024: Coin Metadata
+
+## Changelog
+
+- 05/19/2020: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+Assets in the Cosmos SDK are represented via a `Coins` type that consists of an `amount` and a `denom`,
+where the `amount` can be any arbitrarily large or small value. In addition, the Cosmos SDK uses an
+account-based model where there are two types of primary accounts -- basic accounts and module accounts.
+All account types have a set of balances that are composed of `Coins`. The `x/bank` module keeps
+track of all balances for all accounts and also keeps track of the total supply of balances in an
+application.
+
+With regards to a balance `amount`, the Cosmos SDK assumes a static and fixed unit of denomination,
+regardless of the denomination itself. In other words, clients and apps built atop a Cosmos-SDK-based
+chain may chose to define and use arbitrary units of denomination to provide a richer UX, however, by
+the time a tx or operation reaches the Cosmos SDK state machine, the `amount` is treated as a single
+unit. For example, for the Cosmos Hub (Gaia), clients assume 1 ATOM = 10^6 uatom, and so all txs and
+operations in the Cosmos SDK work off of units of 10^6.
+
+This clearly provides a poor and limited UX especially as interoperability of networks increases and
+as a result the total amount of asset types increases. We propose to have `x/bank` additionally keep
+track of metadata per `denom` in order to help clients, wallet providers, and explorers improve their
+UX and remove the requirement for making any assumptions on the unit of denomination.
+
+## Decision
+
+The `x/bank` module will updated to store and index metadata by `denom`, specifically the "base" or
+smallest unit -- the unit the Cosmos SDK state-machine works with. The metadata should ideally be
+extensible and serve the needs of many external consumers and even other modules (e.g. `x/ibc` or `x/nft`).
+At a minimum, the metadata type should include all units and a potential list of denomination aliases.
+As a result, we can define the type as follows:
+
+```protobuf
+message DenomUnit {
+  string denom    = 1;
+  uint32 exponent = 2;  
+}
+
+message Metadata {
```

Comment:
We should add a `source_chain` field for relayed tokens too. See https://github.com/cosmos/ics/blob/master/spec/ics-020-fungible-token-transfer/README.md#multi-chain-notes (2nd paragraph):

> In order to track all of the denominations moving around the network of chains in various paths, it may be helpful for a particular chain to implement a registry which will track the "global" source chain for each denomination. End-user service providers (such as wallet authors) may want to integrate such a registry or keep their own mapping of canonical source chains and human-readable names in order to improve UX.

Reply:
something we can add too is a boolean to signal if a denomination is allowed to be transferred through IBC. This can be useful for new unstable chains (like the `SendEnabled` param we had in bank) and for private/consortium chains that may want to receive external denoms but not transfer their native tokens out.

Reply:
I don't want to conflate IBC here -- this is meant to be generic coin metadata. I'd rather come up with a more general approach to metadata extensions so that we can add IBC related fields if necessary.

Reply:
I agree that we shouldn't build anything into the coin metadata spec that is specific to IBC.

The registry of source chains would have to be implemented separately, it needs more complex governance.